ccm_web/server/src/canvas/canvas.service.ts

```diff
@@ -47,75 +51,138 @@ export class CanvasService {
     return `${this.url}/login/oauth2/auth?${searchParams.toString()}`
   }
 
-  async createTokenForUser (userLoginId: string, canvasCode: string): Promise<boolean> {
+  static calculateExpiresAt (time: Date, expiresIn: number): Date {
+    // Use 95 percent of expiration period to guarantee expired tokens aren't used
+    const msTilExpire = expiresIn * 1000 * 0.95
+    logger.debug(`Milliseconds until token should be treated as expired: ${msTilExpire}`)
+    const expiresAt = new Date(time.getTime() + msTilExpire)
+    logger.debug(`Treat token as expired at ${expiresAt.toISOString()}`)
+    return expiresAt
+  }
+
+  async createTokenForUser (userLoginId: string, canvasCode: string): Promise<void> {
     /*
     Make a call to the Canvas API to create token
     https://canvas.instructure.com/doc/api/file.oauth_endpoints.html#post-login-oauth2-token
     */
 
+    const user = await this.userService.findUserByLoginId(userLoginId)
+    if (user === null) {
+      logger.error(`User ${userLoginId} is not in the database.`)
+      throw new UserNotFoundError(userLoginId)
+    }
```

Comment:
If/when we implement Passport and attach `User` objects to `Request`s, we will be able to throw this error in one place, and I can instead pass a `User` object in the first parameter position to a few `CanvasService` methods. I will tackle that with the Passport work.

I could get even closer to this by moving the `UserService` calls to the controllers, but I wanted to minimize changes at this point.